### PR TITLE
Adding all possible data type interactions to the perf tests since some use SIMD acceleration and others do not.

### LINKED ISSUE #15439: Not all data type combinations are being tested with the integral function performance tests
##### System information (version)
- OpenCV => 3.4
- Operating System / Platform => Linux 64bit
- Compiler => gcc

##### Detailed description

I'm trying to analyse the performance of the integral function for several parameter
combinations since some of those combinations use HAL instructions and others do
not. Running the performance tests I realised that only one combinations of matrix
size for square sum and sum matrices.

### PR DESCRIPTION
**Pull request by everton1984**
resolves #15439

### This pullrequest changes

This PR is changing modules/imgproc/perf/perf_integral.cpp if you look at sumpixels.cpp there are several possible combinations of matrix sizes for the integral function. The performance tests do not cover all possibilities even with some of them being optimised with HAL instructions so it's impossible to analyse improvements made for those combinations.

### CODE REVIEW

File: modules/imgproc/perf/perf_integral.cpp [hunk 1]
```diff
@@ -5,9 +5,22 @@
...
 
+enum PerfSqMatDepth{
+    CV_32S32S = 0,
+    CV_32S32F,
+    CV_32S64F,
+    CV_32F32F,
+    CV_32F64F,
+    CV_64F64F};
+
+CV_ENUM(SqSizes, CV_32S32S, CV_32S32F, CV_32S64F, CV_32F32F, CV_32F64F, CV_64F64F);
```
Comment (alalek) [inline modules/imgproc/perf/perf_integral.cpp]: > SqSizes

This doesn't look as "sizes". Perhaps it should be `IntergralOutputDepths`

Also it is better to replace `CV_` prefix to avoid confusion with public OpenCV API.

File: modules/imgproc/perf/perf_integral.cpp [hunk 2]
```diff
@@ -31,28 +44,57 @@ PERF_TEST_P(Size_MatType_OutMatDepth, integral,
...
+    SANITY_CHECK(sqsum, 1e-6);
+}
+
+int vals[6][2] = {{CV_32S, CV_32S}, {CV_32S, CV_32F}, {CV_32S, CV_64F}, {CV_32F, CV_32F}, {CV_32F, CV_64F}, {CV_64F, CV_64F}};
+PERF_TEST_P(Size_MatType_OutMatDepthArray, DISABLED_integral_sqsum_full,
             testing::Combine(
                 testing::Values(TYPICAL_MAT_SIZES),
                 testing::Values(CV_8UC1, CV_8UC2, CV_8UC3, CV_8UC4),
-                testing::Values(CV_32S, CV_32F, CV_64F)
+                testing::Values(0,1,2,3,4,5)
```
Comment (alalek) [inline modules/imgproc/perf/perf_integral.cpp]: Avoid using of magic numbers. Use enum constants identifiers instead.

File: modules/imgproc/perf/perf_integral.cpp [hunk 3]
```diff
@@ -31,28 +44,57 @@ PERF_TEST_P(Size_MatType_OutMatDepth, integral,
...
+    Mat sqsum(sz, sqdepth);
 
     declare.in(src, WARMUP_RNG).out(sum, sqsum);
     declare.time(100);
 
-    TEST_CYCLE() integral(src, sum, sqsum, sdepth);
+    TEST_CYCLE() integral(src, sum, sqsum, sdepth, sqdepth);
 
     SANITY_CHECK(sum, 1e-6);
     SANITY_CHECK(sqsum, 1e-6);
```
Comment (alalek) [inline modules/imgproc/perf/perf_integral.cpp]: You don't have sanity data for this test (in opencv_extra repositry). Replace these 2 lines by `SANITY_CHECK_NOTHING();`

Comment (everton1984) [reply modules/imgproc/perf/perf_integral.cpp]: Thanks for the feedback, just made the changes you requested.

File: modules/imgproc/perf/perf_integral.cpp [hunk 4]
```diff
@@ -55,6 +68,35 @@ PERF_TEST_P(Size_MatType_OutMatDepth, integral_sqsum,
...
+    Mat sqsum(sz, sqdepth);
+
+    declare.in(src, WARMUP_RNG).out(sum, sqsum);
+    declare.time(100);
+
+    TEST_CYCLE() integral(src, sum, sqsum, sdepth, sqdepth);
+
+    SANITY_CHECK_NOTHING();
+    SANITY_CHECK_NOTHING();
+
```
Comment (alalek) [inline modules/imgproc/perf/perf_integral.cpp]: One macro should be enough

File: modules/imgproc/perf/perf_integral.cpp [hunk 5]
```diff
@@ -31,12 +44,12 @@ PERF_TEST_P(Size_MatType_OutMatDepth, integral,
...
-                testing::Values(CV_8UC1, CV_8UC2, CV_8UC3, CV_8UC4),
-                testing::Values(CV_32S, CV_32F, CV_64F)
-                )
-            )
+                testing::Combine(
+                    testing::Values(TYPICAL_MAT_SIZES),
+                    testing::Values(CV_8UC1, CV_8UC2, CV_8UC3, CV_8UC4),
+                    testing::Values(CV_32S, CV_32F, CV_64F)
+                    )
+        )
```
Comment (alalek) [inline modules/imgproc/perf/perf_integral.cpp]: Lets avoid unrelated changes.

File: modules/imgproc/perf/perf_integral.cpp [hunk 6]
```diff
@@ -5,9 +5,22 @@
...
 
+enum PerfSqMatDepth{
+    SZ_32S32S = 0,
+    SZ_32S32F,
+    SZ_32S64F,
+    SZ_32F32F,
+    SZ_32F64F,
+    SZ_64F64F};
+
+CV_ENUM(IntegralOutputDepths, SZ_32S32S, SZ_32S32F, SZ_32S64F, SZ_32F32F, SZ_32F64F, SZ_64F64F);
```
Comment (alalek) [inline modules/imgproc/perf/perf_integral.cpp]: > SZ_

What is the meaning of this?
Depth != Size (size is usually associated with image size in pixels, not a "size" of type)

`DEPTH_32S_32S`?

File: modules/imgproc/perf/perf_integral.cpp [hunk 7]
```diff
@@ -55,6 +68,35 @@ PERF_TEST_P(Size_MatType_OutMatDepth, integral_sqsum,
     SANITY_CHECK(sqsum, 1e-6);
 }
 
+int vals[6][2] = {{CV_32S, CV_32S}, {CV_32S, CV_32F}, {CV_32S, CV_64F}, {CV_32F, CV_32F}, {CV_32F, CV_64F}, {CV_64F, CV_64F}};
```
Comment (alalek) [inline modules/imgproc/perf/perf_integral.cpp]: This should be close to related enumeration above.
This should be properly named.

This must be "static" or in anonymous namespace due its "local" usage (we don't want conflicts especially on such common name).

**Comment by everton1984** [comment]
Changed the names per request, made the array static and moved the definition closer to the enum.
